### PR TITLE
Refactory particle system update.

### PR DESCRIPTION
**Pull request by fusijie**
To improve #11566.

The comparison of particle update.

Env:  before = v3.8, after = v3.8 + this pr. iOS(50 ParticleFireworks). Android (30 ParticleFireworks).
eb75da33-520e-47aa-bb62-d6031afb86bb

### CODE REVIEW

**Comment by fusijie** [comment]
@dabingnn please help to review.

**Comment by fusijie** [comment]
[ci rebuild]

File: cocos/2d/CCParticleSystem.h [hunk 1]
```diff
@@ -43,43 +43,101 @@ class ParticleBatchNode;
...
+        float* deltaRadius;
     } modeB;
+    
+    unsigned int maxCount;
+    ParticleData();
+    bool init(int count);
+    void release();
+    inline unsigned int getMaxCount(){ return maxCount;}
+    
+    inline void swapParticle(int p1,int p2)
```
Comment (dabingnn) [inline cocos/2d/CCParticleSystem.h]: It seems that it is not swap.

Comment (fusijie) [inline cocos/2d/CCParticleSystem.h]: I forgot rename it and I'll rename it to `copy`. :-D

File: cocos/2d/CCParticleSystem.cpp [hunk 2]
```diff
@@ -478,134 +581,198 @@ ParticleSystem::~ParticleSystem()
...
+        {
+            _particleData.modeA.tangentialAccel[i] = modeA.tangentialAccel + modeA.tangentialAccelVar * RANDOM_M11(&RANDSEED);
+        }
+        
         // rotation is dir
-        if(modeA.rotationIsDir)
-            particle->rotation = -CC_RADIANS_TO_DEGREES(particle->modeA.dir.getAngle());
+        if( modeA.rotationIsDir )
+        {
+            for (int i = start; i < _particleCount; ++i)
```
Comment (dabingnn) [inline cocos/2d/CCParticleSystem.cpp:714]: this code is repeated in else block, could we refactor it to be more clearer?

File: cocos/2d/CCParticleSystem.cpp [hunk 3]
```diff
@@ -833,10 +883,106 @@ void ParticleSystem::update(float dt)
...
+                if (_particleData.posx[i] || _particleData.posy[i])
+                {
+                    nomalize_point(_particleData.posx[i], _particleData.posy[i], &radial);
+                }
+                tangential = radial;
+                radial.x *= _particleData.modeA.radialAccel[i];
+                radial.y *= _particleData.modeA.radialAccel[i];
+                
+                // tangential acceleration
+                tangential.x = -tangential.y;
```
Comment (dabingnn) [inline cocos/2d/CCParticleSystem.cpp]: there is a bug here, `std::swap`could be used instead.

File: cocos/2d/CCParticleSystem.cpp [hunk 4]
```diff
@@ -833,10 +883,106 @@ void ParticleSystem::update(float dt)
...
+                // if (_configName.length()>0 && _yCoordFlipped != -1)
+                
+                // this is cocos2d-x v3.0
+                tmp.x = _particleData.modeA.dirX[i] * dt * _yCoordFlipped;
+                tmp.y = _particleData.modeA.dirY[i] * dt * _yCoordFlipped;
+                _particleData.posx[i] += tmp.x;
+                _particleData.posy[i] += tmp.y;
+            }
+        }
+        else
```
Comment (dabingnn) [inline cocos/2d/CCParticleSystem.cpp:929]: do not remove comments, which is hard to maintain

File: cocos/2d/CCParticleSystem.cpp [hunk 5]
```diff
@@ -674,158 +841,41 @@ void ParticleSystem::update(float dt)
...
-
-                // angle
-                p->rotation += (p->deltaRotation * dt);
-
-                //
-                // update values in quad
-                //
-
-                Vec2    newPos;
-
```
Comment (dabingnn) [inline cocos/2d/CCParticleSystem.cpp:776]: These code are missed, please confirm that whether it is should be deleted or not.

Comment (fusijie) [inline cocos/2d/CCParticleSystem.cpp:776]: Not missed, these code are removed to ParticleSystemQuad.

File: cocos/2d/CCParticleSystem.cpp (hunk 2 above)
Comment (fusijie) [inline cocos/2d/CCParticleSystem.cpp:714]: I think it's not very necessary, and the abstract method is hard to maintain.

File: cocos/2d/CCParticleSystem.cpp (hunk 3 above)
Comment (fusijie) [inline cocos/2d/CCParticleSystem.cpp]: I'll check it.

File: cocos/2d/CCParticleSystem.cpp (hunk 4 above)
Comment (fusijie) [inline cocos/2d/CCParticleSystem.cpp:929]: I'll restore them.

**Comment by pandamicro** [comment]
@fusijie  Please complete this PR so that QA team can test it

**Pull request merged**